File: modules/umeng-bridge/android/src/main/java/com/west2online/umeng/ExpoUmengModule.kt
package com.west2online.umeng

import android.app.ActivityManager
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.provider.Settings
import android.util.Log
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationManagerCompat
import com.umeng.analytics.MobclickAgent
import com.umeng.commonsdk.UMConfigure
import com.umeng.message.PushAgent
import com.umeng.message.UmengNotificationClickHandler
import com.umeng.message.api.UPushRegisterCallback
import com.umeng.message.entity.UMessage
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext

class ExpoUmengModule : Module() {
    private var upushRegistered = false
    private var error = ""

    class ResponseMapper : Record {
        // tag 集合 是一个字符串数组，默认为空
        @Field
        var data: List<String> = emptyList()

        // 剩余可用的 tag 数
        @Field
        var remain: Int = 0

        // 错误信息
        @Field
        var error: String = ""
    }

    override fun definition() = ModuleDefinition {
        Name("ExpoUmeng")

        Function("initUmeng") {
            CoroutineScope(Dispatchers.IO).launch {
                val metadata = context.packageManager.getApplicationInfo(
                    context.applicationInfo.packageName,
                    PackageManager.GET_META_DATA
                ).metaData
                UMConfigure.init(
                    context,
                    metadata.getString("UMENG_APPKEY"),
                    metadata.getString("UMENG_CHANNEL"),
                    UMConfigure.DEVICE_TYPE_PHONE,
                    metadata.getString("UMENG_MSGSEC")
                )
                PushAgent.getInstance(context).onAppStart()
                PushAgent.getInstance(context)
                    .register(object : UPushRegisterCallback {
                        override fun onSuccess(deviceToken: String) {
                            Log.d("UMLog", "deviceToken:$deviceToken")
                            upushRegistered = true
                        }

                        override fun onFailure(errCode: String, errDesc: String) {
                            Log.e("UMLog", "注册失败 code:$errCode, desc:$errDesc")
                            upushRegistered = false
                            appendError("PushAgent.getInstance(context).register 注册失败 code:$errCode, desc:$errDesc")
                        }
                    })

                MobclickAgent.setPageCollectionMode(MobclickAgent.PageMode.AUTO)
                val notificationClickHandler: UmengNotificationClickHandler =
                    object : UmengNotificationClickHandler() {
                        override fun launchApp(context: Context, msg: UMessage) {
                            val newMsg = msg.extra;
                            Log.i("UMLog", "msg:$newMsg")
                        }
                    }
                PushAgent.getInstance(context).notificationClickHandler = notificationClickHandler

                if (BuildConfig.DEBUG) {
                    withContext(Dispatchers.Main) {
                        Toast.makeText(context, "Umeng初始化流程已走完", Toast.LENGTH_SHORT).show()
                    }
                }
            }
            return@Function null
        }

        Function("hasPermission") {
            // 检查通知权限
            return@Function NotificationManagerCompat.from(context).areNotificationsEnabled()
        }

        Function("requirePermission") {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                // 申请通知权限弹窗
                val activity = appContext.currentActivity
                if (activity != null) {
                    ActivityCompat.requestPermissions(
                        activity,
                        arrayOf(android.Manifest.permission.POST_NOTIFICATIONS),
                        1
                    )
                    CoroutineScope(Dispatchers.Main).launch {
                        // 300ms内弹窗未弹出（App仍在前台），且已拒绝，说明是永久拒绝
                        delay(300)
                        if (isForegroundActivityFromMyApp(activity) &&
                            !NotificationManagerCompat.from(context).areNotificationsEnabled()
                        ) {
                            gotoPermissionPage()
                        }
                    }
                } else {
                    Log.e("UMLog", "Activity is null")
                    gotoPermissionPage()
                }
            } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                gotoPermissionPage()
            } else {
                // granted by default
            }
        }

        Function("getDeviceToken") {
            return@Function PushAgent.getInstance(context).registrationId
        }

        Function("isRegisteredForRemoteNotifications") {
            return@Function upushRegistered
        }

        Function("getError") {
            return@Function error
        }

        Function("getAppKeyAndChannel") {
            val metadata = context.packageManager.getApplicationInfo(
                context.applicationInfo.packageName,
                PackageManager.GET_META_DATA
            ).metaData
            return@Function metadata.getString("UMENG_APPKEY") + ", " +
                    metadata.getString("UMENG_CHANNEL")
        }

        Function("getAllTags") {
            return@Function runBlocking(Dispatchers.IO) {
                val response = ResponseMapper()
                val deferred = CompletableDeferred<ResponseMapper>()
                PushAgent.getInstance(context).tagManager.getTags { isSuccess, data ->
                    if (isSuccess) {
                        response.data = data
                        // 目前每个用户tag限制在1024个
                        // https://developer.umeng.com/docs/67966/detail/98583#h1--tag-alias-4
                        response.remain = 1024 - data.size
                    } else {
                        response.error = "ExpoUmengModule: Failed to get tags"
                    }
                    deferred.complete(response)
                }
                deferred.await()
            }
        }

        Function("addTags") { tags: List<String> ->
            return@Function runBlocking(Dispatchers.IO) {
                val response = ResponseMapper()
                val deferred = CompletableDeferred<ResponseMapper>()
                PushAgent.getInstance(context).tagManager.addTags({ isSuccess, result ->
                    if (isSuccess) {
                        response.data = listOf("ExpoUmengModule: 友盟安卓SDK添加接口不返回tag列表")
                        response.remain = result.remain
                    } else {
                        response.error = "ExpoUmengModule: Failed to add tags"
                    }
                    deferred.complete(response)
                }, *tags.toTypedArray())
                deferred.await()
            }
        }

        Function("deleteTags") { tags: List<String> ->
            return@Function runBlocking(Dispatchers.IO) {
                val response = ResponseMapper()
                val deferred = CompletableDeferred<ResponseMapper>()
                PushAgent.getInstance(context).tagManager.deleteTags({ isSuccess, result ->
                    if (isSuccess) {
                        response.data = listOf("ExpoUmengModule: 友盟安卓SDK删除接口不返回tag列表")
                        response.remain = result.remain
                    } else {
                        response.error = "ExpoUmengModule: Failed to delete tags"
                    }
                    deferred.complete(response)
                }, *tags.toTypedArray())
                deferred.await()
            }
        }
    }

    private val context
        get() = requireNotNull(appContext.reactContext)

    // 跳转权限设置页面
    @RequiresApi(Build.VERSION_CODES.O)
    private fun gotoPermissionPage() {
        val activity = appContext.currentActivity
        val intent = Intent(Settings.ACTION_APP_NOTIFICATION_SETTINGS).apply {
            putExtra(Settings.EXTRA_APP_PACKAGE, context.packageName)
        }
        if (activity != null) {
            activity.startActivity(intent)
        } else {
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            context.startActivity(intent)
        }
    }

    private fun isForegroundActivityFromMyApp(context: Context): Boolean {
        try {
            // 获取 ActivityManager 实例
            val activityManager =
                context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager

            @Suppress("DEPRECATION") val tasks = activityManager.getRunningTasks(1)

            if (tasks.isNotEmpty()) {
                // 获取前台 Activity 的信息
                val topActivity = tasks[0].topActivity

                // 检查前台 Activity 是否属于本应用
                if (topActivity != null && topActivity.packageName == context.packageName) {
                    return true
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return false
    }

    private fun appendError(err: String) {
        error += err + "\n\n"
    }
}

File: modules/umeng-bridge/ios/ExpoUmengModule.swift
import ExpoModulesCore
import UserNotifications
import UIKit

// 用于统一返回数据格式
struct ResponseMapper: Record {
  // tag 集合 是一个字符串数组，默认为空
  @Field var data = [String]()
  // 剩余可用的 tag 数
  @Field var remain = 0
  // 错误信息
  @Field var error = ""
}

public class ExpoUmengModule: Module {
    public func definition() -> ModuleDefinition {
        Name("ExpoUmeng")

        // 初始化友盟推送
        Function("initUmeng") { () -> Void in
          PushNotificationManager.shared.setAllowPrivacy()
          // 在 Manager 中，允许隐私政策后会自动初始化推送
          // 这个设置会写入 iOS 原生存储 UserDefaults 中
          // 下次开 app 时，会在 AppDelegate 中标准的初始化入口自动调用初始化推送
          // 既合规（要允许隐私政策才能初始化）也符合生命周期原则
        }

        // 检查通知权限
        Function("hasPermission") { () -> Bool in
            var hasPermission = false
            let semaphore = DispatchSemaphore(value: 0)

            UNUserNotificationCenter.current().getNotificationSettings { settings in
                hasPermission = settings.authorizationStatus == .authorized
                semaphore.signal()
            }

            _ = semaphore.wait(timeout: .now() + 5) // 等待权限检查
            return hasPermission
        }

        // 请求通知权限，待删除
        Function("requirePermission") { () -> Void in
          // iOS 不需要这个接口，现在是打开 App 时就会请求通知权限
        }

        // 返回设备的 deviceToken
        Function("getDeviceToken") { () -> String in
            return PushNotificationManager.shared.getDeviceToken()
        }

        // 返回设备的远程推送注册状态
        Function("isRegisteredForRemoteNotifications") { () -> Bool in
            return PushNotificationManager.shared.isRegisteredForRemoteNotifications()
        }

        // 返回错误信息
        Function("getError") { () -> String in
            return PushNotificationManager.shared.getError()
        }

        // 获取 AppKey 以及 Channel
        Function("getAppKeyAndChannel") { () -> String in
          guard let infoDictionary = Bundle.main.infoDictionary else {
              return "没有找到 Info.plist"
          }
          let appKey = infoDictionary["UMENG_APPKEY"] as? String ?? ""
          let channel = infoDictionary["UMENG_CHANNEL"] as? String ?? ""
          return "\(appKey), \(channel)"
        }

        // 获取所有 tag
        Function("getAllTags") { () -> ResponseMapper in
            var resp = ResponseMapper(data: [], remain: 0, error: "")
            let semaphore = DispatchSemaphore(value: 0) // 使用 DispatchSemaphore 来同步等待异步操作完成
            UMPushSwift.getTags { tags, remain, error in
                // 将 tags 转换为 Swift 的 Set<String>，然后再转换为 [String]
                if let tags = tags as? Set<String> {
                    resp.data = Array(tags)
                } else if let tags = tags as? NSSet {
                    resp.data = tags.allObjects as? [String] ?? []
                } else {
                    resp.data = []
                }
                resp.remain = remain
                resp.error = error?.localizedDescription ?? ""

                semaphore.signal()
            }
            semaphore.wait() // 等待异步操作完成
            return resp
        }

        // 添加 tag
        Function("addTags") { (tag: [String]) in
            var resp = ResponseMapper(data: [], remain: 0, error: "")
            let semaphore = DispatchSemaphore(value: 0) // 使用 DispatchSemaphore 来同步等待异步操作完成
            UMPushSwift.addTags(tag) { tags, remain, error in
                // 将 tags 转换为 Swift 的 Set<String>，然后再转换为 [String]
                if let tags = tags as? Set<String> {
                    resp.data = Array(tags)
                } else if let tags = tags as? NSSet {
                    resp.data = tags.allObjects as? [String] ?? []
                } else {
                    resp.data = []
                }
                resp.remain = remain
                resp.error = error?.localizedDescription ?? ""
                semaphore.signal()
            }
            semaphore.wait() // 等待异步操作完成
        }

        // 删除 tag
        Function("deleteTags") { (tag: [String]) in
            var resp = ResponseMapper(data: [], remain: 0, error: "")
            let semaphore = DispatchSemaphore(value: 0) // 使用 DispatchSemaphore 来同步等待异步操作完成
            UMPushSwift.deleteTags(tag) { tags, remain, error in
                // 将 tags 转换为 Swift 的 Set<String>，然后再转换为 [String]
                if let tags = tags as? Set<String> {
                    resp.data = Array(tags)
                } else if let tags = tags as? NSSet {
                    resp.data = tags.allObjects as? [String] ?? []
                } else {
                    resp.data = []
                }
                resp.remain = remain
                resp.error = error?.localizedDescription ?? ""
                semaphore.signal()
            }
            semaphore.wait() // 等待异步操作完成
        }
    }
}
